### DISCUSSION TITLE
Understanding LocalNormalizedCrossCorrelationLoss

### DISCUSSION THREAD

**Discussion by ArjunNarayanan** [Project-MONAI/MONAI#6645] [Q&A] [answered]
Hello,

I'm trying to understand how to use `monai.losses.LocalNormalizedCrossCorrelationLoss` for image registration. I'm finding it a little difficult to understand how to interpret this loss. For example, if I take two images and compute the LNCC loss,

```
from monai.losses import LocalNormalizedCrossCorrelationLoss
lncc_loss = LocalNormalizedCrossCorrelationLoss(kernel_size=3)
example_image # input image of size [B,C,128,128,128] 

self_loss = lncc_loss(example_image, example_image) # result is tensor(-0.1860) for my image
```

I would expect the cross-correlation of an image with itself to be $-1$ but this is not the case. When I make the `kernel_size=127` (which corresponds to NCC instead of LNCC if my understanding is correct), the loss value is indeed $-1$. Can someone explain this to me?

**Comment by wyli** [comment]
thanks, I've tested this example will return -1...
```py
import torch

from monai.losses import LocalNormalizedCrossCorrelationLoss

lncc_loss = LocalNormalizedCrossCorrelationLoss(kernel_size=3)
example_image = torch.randn(3, 2, 128, 128, 128, dtype=float)

self_loss = lncc_loss(example_image, example_image)
print(self_loss)
```

but `example_image = torch.ones(3, 2, 128, 128, 128, dtype=float)` doesn't return -1 because the small constant added https://github.com/Project-MONAI/MONAI/blob/6de86a4f3931cd1cadeb292a32ccd16b0e5c5d70/monai/losses/image_dissimilarity.py#L70-L71

**Comment by ArjunNarayanan** [comment, marked as answer]
Okay that makes sense. Since an image with all constants will have a zero variance, the LNCC will be undefined so we have to add a small smoothing constant. In my case, the image had a lot of background pixels that were constant and so this was ending up affecting the LNCC. If I compute the loss with `reduction="none"` and mask out the background pixels, the LNCC is indeed $-1$. Thanks for your help in understanding this :)
